pkg/storegateway/partitioner.go

```diff
@@ -82,12 +86,23 @@ func (g *gapBasedPartitioner) partition(length int, rng func(int) (uint64, uint6
 
 		p := Part{}
 		p.Start, p.End = rng(j)
+		stats.requestedRangesTotal += p.End - p.Start
 
 		// Keep growing the range until the end or we encounter a large gap.
 		for ; k < length; k++ {
 			s, e := rng(k)
 
-			if p.End+g.maxGapBytes < s {
+			if p.End >= s {
+				// The start of the next range overlaps with the current range's end, so we can merge them.
+				// We count the extra bytes between the current range's end and the next one's end - that's what's been requested.
+				stats.requestedRangesTotal += e - p.End
```

Comment:
Isn't there a bug in the case of the ranges `[10, 20]`, `[15, 18]`? I haven't tried to execute the code with this case, but I think we subtract for the 2 range instead of "adding zero".

Reply:
nice catch. I will add a test case for this

Reply:
added one in https://github.com/grafana/mimir/pull/3769/commits/f646b87e7c6f8c0034f7fd14dbe0dd4dc4b7aede